data_structures.md

```diff
@@ -11,12 +11,601 @@ kernelspec:
   name: python3
 ---
 
+```{currentmodule} tskit
+ ```
+
 (sec_data_structures)=
 
 # Data structures
 
+:::{todo}
+The content here has been ported from the msprime tutorial via the tskit tutorial and
+needs to be reorganised to make a coherent narrative.
+:::
+
+(sec_data_structures_tables)=
 
 ## Tables
 
+Internally, a tree sequence can be thought of as a set of tables, and {program}`tskit`
+provides an interface for dealing with these tables directly. This is particularly
+relevant when you wish to {ref}`edit or otherwise modify<sec_data_structures_editing>`
+a tree sequence.
+
+There are eight tables that together define a tree sequence, although some may be empty,
+and together they form a {class}`TableCollection`.
+The tables are defined in the official {program}`tskit` documentation for
+{ref}`Table Definitions <tskit:sec_table_definitions>`, and the
+{ref}`Tables API <tskit:sec_tables_api>` section in the docs describes how to work with
+them. Here we make some general remarks about what you can and cannot do with them.
+
+{program}`tskit` provides direct access to the columns of each table as
+{program}`numpy` arrays: for instance, if ``n`` is a {class}`NodeTable`, then ``n.time``
+will return an array containing the birth times of the individuals whose genomes
+are represented by the nodes in the table.
+*However*, it is important to note that this is *not* a shallow copy:
+modifying ``n.time`` will *not* change the node table ``n``.  This may change in
+the future, but currently there are three ways to modify tables: ``.add_row()``,
+``.set_columns()``, and ``.append_columns()``
+(and also ``.clear()``, which empties the table).
+
+:::{todo}
+Update this to reflect the new slice methods for accessing and setting rows.
+:::
+
+For example, a node table could be constructed using {meth}`NodeTable.add_row` as
+follows:
+
+```{code-cell} ipython3
+import tskit
+
+n = tskit.NodeTable()
+sv = [True, True, True, False, False, False, False]
+tv = [0.0, 0.0, 0.0, 0.4, 0.5, 0.7, 1.0]
+pv = [0, 0, 0, 0, 0, 0, 0]
+for s, t, p in zip(sv, tv, pv):
+    n.add_row(flags=s, population=p, time=t)
+n
+```
+
+
+The ``.add_row()`` method is natural (and should be reasonably efficient) if
+new records appear one-by-one. In the example above it would have been more
+natural to use ``.set_columns()`` --- equivalently:
+
+```{code-cell} ipython3
+n = tskit.NodeTable()
+n.set_columns(flags=sv, population=pv, time=tv)
+```
+
+Since columns cannot be modified directly as properties of the tables,
+they must be extracted, modified, then replaced.
+:::{todo}
```

Comment:
I don't think it's true anyway, you can do things like ``n.flags = sv``, can't you?

Reply:
Probably easier to put in an idiomatic example here now rather than update the TODO

Reply:
@benjeffery - can you work out a small bit of replacement text for this bit, as you know what the new getter and setter methods are and what the correct idiom is?

Reply:
I think it's literally ``table.time = time_array`` will do it. This calls set_columns under the hood

Reply:
Oh, right, I thought we might want to showcase the slicing operators

Reply:
OK, but that's a separate thing. I just don't want to put in code here that's telling people to do things that we actively want to discourage